keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -111,153 +137,271 @@ Following conditions must be met:
-    * Especially for shared filesystems, it's task of a CSI driver to:
-      * Detect `context` mount option and add any other mount option that allow mount options that's needed to allow mounting one volume with multiple SELinux contexts on the same node.
-        E.g. add `nosharesock` to `mount -t cifs -o context=XYZ`.
-      * Either not support `NodeStageVolume` CSI call or use different VolumeHandle for each PV, even though they lead to the same volume in the storage backend.
-        * This is because kubelet / VolumeManager can track only one global volume mount (created by `NodeStageVolume`) and that mount can have only one SELinux context.
-          Volumes with different VolumeHandles have different global mounts and therefore can have different SELinux contexts.
-  * kubelet does not pass any special SELinux option to the container runtime (explicitly, no `:Z`).
-    Files on the volume already have the right SELinux context, no recursive relabeling happens there.
-  * When a `PodSecurityContext` specifies incomplete SELinux label (i.e. omits `SELinuxOptions.User`, `.Role` or `.Type`), kubelet fills the blanks from the system defaults provided by [ContainerLabels() from go-selinux bindings](https://github.com/opencontainers/selinux/blob/621ca21a5218df44259bf5d7a6ee70d720a661d5/go-selinux/selinux_linux.go#L770).
+* Allow CSI drivers to announce if they support mounting with `-o context` mount option via `CSIDriver spec.SELinuxMount bool` field.
+  * Under `SELinuxMountReadWriteOncePod` feature gate, beta + enabled by default in 1.28.
+  * Default value is `false`, i.e. no support for SELinux mount. It needs explicit opt-in.
+
+
+* Update kubelet to mount all volumes with `-o context` mount option, unless a Pod explicitly opts out.
+  * **Phase 1**: limited to ReadWriteOncePod (RWOP) volumes, under `SELinuxMountReadWriteOncePod` feature gate. RWOP volumes can't be used by multiple Pods and thus this cannot break any workload that mixes usage of the same volume by multiple Pods in parallel.
+  * **Phase 2**: same as Phase 1, the feature is still limited to RWOP volumes.
+    This phase enables opt-out that becomes active in Phase 3.
+  * **Phase 3**: all volumes are mounted, under `SELinuxMount` feature gate. It can break existing applications that need to mix privileged and unprivileged Pods using the same volume in parallel.
+    * We propose to send Pod events to immediately show why such Pods are stuck `ContainerCreating`.
+    * We propose metrics to identify such Pods before the `SELinuxMount` feature gate is enabled, to identify potential issues before a cluster upgrade / enabling the feature gate.
+    * We propose metrics to identify such Pods _after_ the `SELinuxMount` feature gate is enabled, to identify pods that are running only because they run on different nodes. If they landed on the same node, one of them would be stuck `ContainerCreating`, because they mix privileged and unprivileged pods.
```

Comment:
An event on pods that may get stuck is also a good idea to help workload authors.

Reply:
If we want events, we need a controller that sends them. kube-state-metrics used to emit metrics does not have any complex logic, all conflict detection would be done by promql queries.